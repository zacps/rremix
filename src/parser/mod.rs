use pest_derive::*;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RemixParser;

#[cfg(test)]
mod tests {
    use crate::format_pair;
    use insta::assert_snapshot;
    use pest::Parser;
    use std::fs;
    use test_case::test_case;

    use super::*;

    #[test_case("../src/standard-lib.rem")]
    //#[test_case("ex/factorial.rem")]
    //#[test_case("ex/factorial2.rem")]
    //#[test_case("ex/primes.rem")]
    #[test_case("ex/test1.rem")]
    #[test_case("ex/test2.rem")]
    #[test_case("ex/test3.rem")]
    #[test_case("ex/test4.rem")]
    #[test_case("ex/test5.rem")]
    #[test_case("ex/test6.rem")]
    #[test_case("ex/test7.rem")]
    #[test_case("ex/test8.rem")]
    #[test_case("ex/test9.rem")]
    #[test_case("ex/test10.rem")]
    #[test_case("ex/test11.rem")]
    #[test_case("ex/test12.rem")]
    #[test_case("ex/test13.rem")]
    #[test_case("ex/test14.rem")]
    #[test_case("ex/test15.rem")]
    #[test_case("ex/test16.rem")]
    #[test_case("ex/test17.rem")]
    #[test_case("ex/test18.rem")]
    #[test_case("ex/test19.rem")]
    #[test_case("ex/test20.rem")]
    #[test_case("ex/test21.rem")]
    #[test_case("ex/test22.rem")]
    #[test_case("ex/test23.rem")]
    #[test_case("ex/test24.rem")]
    #[test_case("ex/test25.rem")]
    #[test_case("ex/test26.rem")]
    // #[test_case("ex/test27.rem")]
    #[test_case("ex/test28.rem")]
    //#[test_case("gr/drawing.rem")]
    //#[test_case("gr/arrows.rem")]
    //#[test_case("gr/squares.rem")]
    //#[test_case("gr/bounce.rem")]
    //#[test_case("ob/obj-demo.rem")]
    //#[test_case("ob/sidewindermaze.rem")]
    //#[test_case("ex/test29.rem")]
    //#[test_case("ex/test30.rem")]
    //#[test_case("ex/test31.rem")]
    //#[test_case("ex/test32.rem")]
    //#[test_case("ex/test33.rem")]
    //#[test_case("ex/test34.rem")]
    //#[test_case("ex/test35.rem")]
    //#[test_case("ex/test36.rem")]
    //#[test_case("ex/test37.rem")]
    //#[test_case("ex/test38.rem")]
    //#[test_case("ex/test39.rem")]
    fn test_parser(path: &'static str) {
        let file =
            fs::read_to_string("tests/".to_owned() + path).expect("could not find test file");
        match RemixParser::parse(Rule::program, &file) {
            Ok(mut parse) => {
                let pair = parse.next();
                assert_snapshot!(path, format_pair(&pair.unwrap(), 0, true))
            }
            Err(e) => {
                eprintln!("{e}");
                panic!("Failed to parse file {path}")
            }
        }
    }
}
