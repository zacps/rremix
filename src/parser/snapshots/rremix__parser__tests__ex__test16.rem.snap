---
source: src/parser/mod.rs
assertion_line: 72
expression: "format_pair(pair.unwrap(), 0, true)"
---
- program
  - eol_comment: "Filtering lists.\r\n"
  - sigiled_comment: "================\r\n"
  - statement > assignment_statement
    - variable: "list"
    - colon: ":"
    - expression > unary_expression > simple_expression > function_call
      - single_name_part: "filter"
      - expression > unary_expression > simple_expression > function_call
        - literal > number: "1"
        - single_name_part: "to"
        - literal > number: "100"
      - single_name_part: "by"
      - expression > unary_expression > simple_expression > variable: "x"
      - single_name_part: "where"
      - continue_indent: ""
      - newindent_block
        - new_indent: "\t"
        - statement > expression > unary_expression > simple_expression > function_call
          - expression > unary_expression > simple_expression > variable: "x"
          - single_name_part: "is"
          - single_name_part: "divisible"
          - single_name_part: "by"
          - literal > number: "5"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "list"
    - single_name_part: "as"
    - single_name_part: "list"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "x"
  - function_definition
    - function_signature
      - name_paren > paren > single_name_part: "number"
      - name_paren > name_part > single_name_part: "is"
      - name_paren > name_part > single_name_part: "not"
      - name_paren > name_part > single_name_part: "divisible"
      - name_paren > name_part > single_name_part: "by"
      - name_paren > name_part > single_name_part: "any"
      - name_paren > name_part > single_name_part: "in"
      - name_paren > paren > single_name_part: "list"
    - colon: ":"
    - function_statements
      - new_indent: "\t"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "for"
        - single_name_part: "each"
        - expression > unary_expression > simple_expression > variable: "divisor"
        - single_name_part: "in"
        - expression > unary_expression > simple_expression > variable: "list"
        - continue_indent: "\t"
        - newindent_block
          - new_indent: "\t"
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "show"
            - literal > list
              - list_item > expression > unary_expression > simple_expression > variable: "divisor"
              - list_item > expression > unary_expression > simple_expression > literal > string: "\" \""
          - continue_indent: "\t\t"
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "if"
            - expression > unary_expression > simple_expression > function_call
              - expression > unary_expression > simple_expression > variable: "number"
              - single_name_part: "is"
              - single_name_part: "divisible"
              - single_name_part: "by"
              - expression > unary_expression > simple_expression > variable: "divisor"
            - continue_indent: "\t\t"
            - newindent_block
              - new_indent: "\t"
              - statement > return_statement > expression > unary_expression > simple_expression > literal > boolean: "false"
          - continue_indent: "\t\t"
          - statement > expression > unary_expression > simple_expression > literal > boolean: "true"
  - eol_comment: "Primes to 100.\r\n"
  - sigiled_comment: "==============\r\n"
  - statement > assignment_statement
    - variable: "primes"
    - colon: ":"
    - expression > unary_expression > simple_expression > literal > list
      - list_item > expression > unary_expression > simple_expression > literal > number: "2"
      - list_item > expression > unary_expression > simple_expression > literal > number: "3"
      - list_item > expression > unary_expression > simple_expression > literal > number: "5"
      - list_item > expression > unary_expression > simple_expression > literal > number: "7"
  - statement > assignment_statement
    - variable: "list"
    - colon: ":"
    - expression > unary_expression > simple_expression > function_call
      - single_name_part: "filter"
      - expression > unary_expression > simple_expression > function_call
        - literal > number: "8"
        - single_name_part: "to"
        - literal > number: "100"
      - single_name_part: "by"
      - expression > unary_expression > simple_expression > variable: "x"
      - single_name_part: "where"
      - continue_indent: ""
      - newindent_block
        - new_indent: "\t"
        - statement > expression > unary_expression > simple_expression > function_call
          - expression > unary_expression > simple_expression > variable: "x"
          - single_name_part: "is"
          - single_name_part: "not"
          - single_name_part: "divisible"
          - single_name_part: "by"
          - single_name_part: "any"
          - single_name_part: "in"
          - expression > unary_expression > simple_expression > variable: "primes"
  - statement > assignment_statement
    - variable: "primes"
    - colon: ":"
    - expression > unary_expression > simple_expression > function_call
      - single_name_part: "join"
      - expression > unary_expression > simple_expression > variable: "primes"
      - single_name_part: "and"
      - expression > unary_expression > simple_expression > variable: "list"
  - sigiled_comment: ". without the newline this crashes with an access violation\r\n"
  - sigiled_comment: ". when run in the combined test program\r\n"
  - sigiled_comment: ".new line\r\n"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "primes"
    - single_name_part: "as"
    - single_name_part: "list"
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "showline"
    - expression > unary_expression > simple_expression > variable: "list"
    - single_name_part: "as"
    - single_name_part: "list"
  - EOI: ""
